File: app/src/main/java/com/rafaelmallare/terraphage/MainActivity.java
package com.rafaelmallare.terraphage;

import android.animation.LayoutTransition;
import android.content.Intent;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v7.widget.CardView;
import android.view.View;
import android.support.design.widget.NavigationView;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.nostra13.universalimageloader.core.ImageLoaderConfiguration;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

import static android.R.attr.id;
import static com.rafaelmallare.terraphage.AttributeType.ATK;
import static com.rafaelmallare.terraphage.AttributeType.DEF;
import static com.rafaelmallare.terraphage.AttributeType.HP;
import static com.rafaelmallare.terraphage.AttributeType.INIT;
import static com.rafaelmallare.terraphage.AttributeType.MDMG;
import static com.rafaelmallare.terraphage.AttributeType.RDMG;
import static com.rafaelmallare.terraphage.AttributeType.SPD;
import static com.rafaelmallare.terraphage.StatType.CHR;
import static com.rafaelmallare.terraphage.StatType.CON;
import static com.rafaelmallare.terraphage.StatType.DEX;
import static com.rafaelmallare.terraphage.StatType.INT;
import static com.rafaelmallare.terraphage.StatType.PER;
import static com.rafaelmallare.terraphage.StatType.STR;
import static com.rafaelmallare.terraphage.Weapon.Rank.D;
import static com.rafaelmallare.terraphage.WeaponType.Melee;
import static com.rafaelmallare.terraphage.WeaponType.Ranged;

public class MainActivity extends AppCompatActivity
        implements NavigationView.OnNavigationItemSelectedListener {
    com.nostra13.universalimageloader.core.ImageLoader imageLoader = com.nostra13.universalimageloader.core.ImageLoader.getInstance();

    static final int REQUEST_CODE_PICKER = 100;

    float scale;

    //region -REGION: View Bindings-
    @BindView(R.id.content_main) RelativeLayout rootView;

    @BindView(R.id.character_name) TextView characterName_view;
    @BindView(R.id.character_path) TextView characterPath_view;

    @BindView(R.id.hp_current_val) TextView hpCurrentVal_view;
    @BindView(R.id.hp_total_val) TextView hpTotalVal_view;
    @BindView(R.id.exp_current_val) TextView expCurrentVal_view;
    @BindView(R.id.exp_total_val) TextView expTotalVal_view;

    @BindView(R.id.init_val) TextView initVal_view;
    @BindView(R.id.spd_val) TextView spdVal_view;
    @BindView(R.id.dmg_val) TextView dmgVal_view;
    @BindView(R.id.def_val) TextView defVal_view;

    @BindView(R.id.per_val) TextView perVal_view;
    @BindView(R.id.chr_val) TextView chrVal_view;
    @BindView(R.id.int_val) TextView intVal_view;
    @BindView(R.id.con_val) TextView conVal_view;
    @BindView(R.id.str_val) TextView strVal_view;
    @BindView(R.id.dex_val) TextView dexVal_view;
    @BindView(R.id.header_image) ImageView headerImage_view;

    @BindView(R.id.atk_card) CardView atkCardView;
    @BindView(R.id.atk_expandable) LinearLayout atkExpandable_view;
    @BindView(R.id.btn_atk_expand) ImageButton btnAtkExpand;
    @BindView(R.id.list_weapons) ListView weaponListView;

    @BindView(R.id.def_card) CardView defCardView;
    @BindView(R.id.def_expandable) LinearLayout defExpandable_view;
    @BindView(R.id.btn_def_expand) ImageButton btnDefExpand;
    //endregion

    final Character character = Character.getInstance();
    final GearStore gearStore = GearStore.getInstance();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        //region -REGION: Initialization stuff (includes FAB code)-
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        ButterKnife.bind(this);

        scale = getResources().getDisplayMetrics().density;

        /***Animation stuff**/
        //LayoutTransition layoutTransition = rootView.getLayoutTransition();
        //layoutTransition.enableTransitionType(LayoutTransition.CHANGING);

        ImageLoaderConfiguration imageLoaderConfig = new ImageLoaderConfiguration.Builder(this)
                .build();

        com.nostra13.universalimageloader.core.ImageLoader.getInstance().init(imageLoaderConfig);

        FloatingActionButton fab = (FloatingActionButton) findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG)
                        .setAction("Action", null).show();
            }
        });

        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(
                this, drawer, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        drawer.setDrawerListener(toggle);
        toggle.syncState();

        NavigationView navigationView = (NavigationView) findViewById(R.id.nav_view);
        navigationView.setNavigationItemSelectedListener(this);
        //endregion

        character.addGear(new Weapon("Revolver", 200, Ranged, ATK, -1, RDMG, 10));
        character.addGear(new Weapon("Great Sword", 15, Melee, ATK, -3, MDMG, 4));
        character.addGear(new Weapon("Flail", 10, Melee, ATK, -3, MDMG, 4));

        updateViewValues();
    }

    public void updateViewValues(){
        characterName_view.setText(character.getName());

        hpTotalVal_view.setText("" + character.getAttribute(HP).getTotalValue());
        hpCurrentVal_view.setText(hpTotalVal_view.getText());

        initVal_view.setText("" + character.getAttribute(INIT).getTotalValue());
        spdVal_view.setText("" + character.getAttribute(SPD).getTotalValue());
        dmgVal_view.setText("" + character.getAttribute(MDMG).getTotalValue());
        defVal_view.setText("" + character.getAttribute(DEF).getTotalValue());

        expTotalVal_view.setText("" + character.getExp());
        expCurrentVal_view.setText(expTotalVal_view.getText());

        perVal_view.setText("" + character.getStat(PER));
        chrVal_view.setText("" + character.getStat(CHR));
        intVal_view.setText("" + character.getStat(INT));
        conVal_view.setText("" + character.getStat(CON));
        strVal_view.setText("" + character.getStat(STR));
        dexVal_view.setText("" + character.getStat(DEX));
    }

    /******test methods*******/
    @OnClick(R.id.dex_val)
    public void increaseDex(){
        character.increaseStat(DEX, 1);
        updateViewValues();
    }

    @OnClick(R.id.str_val)
    public void increaseStr(){
        character.increaseStat(STR, 1);
        updateViewValues();
    }

    @OnClick(R.id.con_val)
    public void increaseCon(){
        character.increaseStat(CON, 1);
        updateViewValues();
    }
    /************************/

    @OnClick({R.id.btn_atk_expand, R.id.btn_def_expand})
    public void expand(ImageButton imgBtn){
        int currentVisibility;
        int newVisibility;
        int pad;
        int id = imgBtn.getId();

        CardView cardView = null;
        LinearLayout expandView = null;
        int backgroundColor;
        int tintColor;
        int expandArrow;

        WeaponListAdapter weaponListAdapter = new WeaponListAdapter(this, character.getWeapons());
        weaponListView.setAdapter(weaponListAdapter);

        if(id == btnAtkExpand.getId()){
            cardView = atkCardView;
            expandView = atkExpandable_view;
        } else if(id == btnDefExpand.getId()){
            cardView = defCardView;
            expandView = defExpandable_view;
        }

        currentVisibility = expandView.getVisibility();

        LinearLayout.LayoutParams params = (LinearLayout.LayoutParams)cardView.getLayoutParams();

        if (currentVisibility == View.GONE){
            params.weight = 1000;
            newVisibility = View.VISIBLE;
            pad = toDP(35);
            backgroundColor = R.color.clear;
            tintColor = R.color.colorExpandableTintDark;
            expandArrow = R.drawable.expand_up;
        } else{
            params.weight = 1;
            newVisibility = View.GONE;
            pad = toDP(15);
            backgroundColor = R.color.colorExpandableBackgroundDark;
            tintColor = R.color.colorExpandableTintLight;
            expandArrow = R.drawable.expand_down;
        }

        imgBtn.setPadding(pad, pad, pad, pad);
        imgBtn.setBackgroundColor(getResources().getColor(backgroundColor));
        imgBtn.setColorFilter(getResources().getColor(tintColor));
        imgBtn.setImageDrawable(getResources().getDrawable(expandArrow));
        cardView.setLayoutParams(params);
        expandView.setVisibility(newVisibility);
    }

    public int toDP(int num){
        return (int) (num * scale + 0.5f);
    }

    @Override
    public void onBackPressed() {
        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START);
        } else {
            super.onBackPressed();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        } else if(id == R.id.action_change_header_image){
            Intent headerImagePicker = new Intent(Intent.ACTION_PICK);
            headerImagePicker.setType("image/*");
            this.startActivityForResult(headerImagePicker, REQUEST_CODE_PICKER);
        }

        return super.onOptionsItemSelected(item);
    }

    @SuppressWarnings("StatementWithEmptyBody")
    @Override
    public boolean onNavigationItemSelected(MenuItem item) {
        // Handle navigation view item clicks here.
        int id = item.getItemId();

        if (id == R.id.nav_camera) {
            // Handle the camera action
        } else if (id == R.id.nav_gallery) {

        } else if (id == R.id.nav_slideshow) {

        } else if (id == R.id.nav_manage) {

        } else if (id == R.id.nav_share) {

        } else if (id == R.id.nav_send) {

        }

        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        drawer.closeDrawer(GravityCompat.START);
        return true;
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == REQUEST_CODE_PICKER && resultCode == RESULT_OK && data != null) {
            imageLoader.displayImage(data.getData().toString(), headerImage_view);
        }
    }
}

File: app/src/main/java/com/rafaelmallare/terraphage/MainActivity.kt
package com.rafaelmallare.terraphage

/**
 * Created by Rj on 5/24/2017.
 */
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.design.widget.NavigationView
import android.support.v4.view.GravityCompat
import android.support.v7.app.ActionBarDrawerToggle
import android.support.v7.app.AppCompatActivity
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ImageButton
import android.widget.LinearLayout
import com.rafaelmallare.terraphage.AttributeType.*
import com.rafaelmallare.terraphage.StatType.*
import com.rafaelmallare.terraphage.WeaponType.*
import com.rafaelmallare.terraphage.GearType.*
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.app_bar_main.*
import kotlinx.android.synthetic.main.content_main.*

class MainActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener {
    val character = Character
    var density = 0f

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(toolbar)

        fab.setOnClickListener { view ->
            Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG)
                    .setAction("Action", null).show()
        }

        val toggle = ActionBarDrawerToggle(
                this, drawer_layout, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close)
        drawer_layout.setDrawerListener(toggle)
        toggle.syncState()

        nav_view.setNavigationItemSelectedListener(this)

        density = resources.displayMetrics.density

        character.inventory.add(Gear("Great Sword", Weapon, Melee, 15, mapOf(ATK to -3, MDMG to 4)))
        character.inventory.add(Gear("Shield", Weapon, Melee, 8, mapOf(ARM to 1)))
        character.inventory.add(Gear("Revolver", Weapon, Ranged, 200, mapOf(ATK to -1, RDMG to 10), 10, 6))
        character.inventory.add(Gear("Leather Armor", Armor, null, 10, mapOf(ARM to 2)))
        character.inventory.add(Gear("Combat Armor", Armor, null, 450, mapOf(ARM to 6, DEF to -1)))
        character.inventory.add(Gear("Dragon Skin Armor", Armor, null, 850, mapOf(ARM to 9, DEF to -3)))

        updateViewValues()

        setOnClickListeners(listOf(btn_atk_expand, btn_def_expand), {
            view ->
                val imgBtn = view as ImageButton
                val expandView: LinearLayout
                val cardView = if (imgBtn.id == btn_atk_expand.id) {
                    list_weapons.adapter = GearListAdapter(this, character.weaponInv, { updateViewValues() })
                    expandView = atk_expandable
                    atk_card
                } else {
                    list_armor.adapter = GearListAdapter(this, character.armorInv, { updateViewValues() })
                    expandView = def_expandable
                    def_card
                }

                val params = cardView.layoutParams as LinearLayout.LayoutParams

                if (expandView.visibility == View.GONE) {
                    params.weight = 1000f
                    imgBtn.setPadding(35.toDP(), 35.toDP(), 35.toDP(), 35.toDP())
                    imgBtn.setBackgroundColor(resources.getColor(R.color.clear))
                    imgBtn.setColorFilter(resources.getColor(R.color.colorExpandableTintDark))
                    imgBtn.setImageDrawable(resources.getDrawable(R.drawable.expand_up))
                    expandView.visibility = View.VISIBLE
                } else {
                    params.weight = 1f
                    imgBtn.setPadding(15.toDP(), 15.toDP(), 15.toDP(), 15.toDP())
                    imgBtn.setBackgroundColor(resources.getColor(R.color.colorExpandableBackgroundDark))
                    imgBtn.setColorFilter(resources.getColor(R.color.colorExpandableTintLight))
                    imgBtn.setImageDrawable(resources.getDrawable(R.drawable.expand_down))
                    expandView.visibility = View.GONE
                }

        })

        //-------------------------------------Test Functions----------------------------------------------
        setOnClickListeners(listOf(per_val, chr_val, int_val, con_val, str_val, dex_val), {
            view -> kotlin.run {
                when(view.id) {
                    per_val.id -> character.increaseStatBy(PER, 1)
                    chr_val.id -> character.increaseStatBy(CHR, 1)
                    int_val.id -> character.increaseStatBy(INT, 1)
                    con_val.id -> character.increaseStatBy(CON, 1)
                    str_val.id -> character.increaseStatBy(STR, 1)
                    dex_val.id -> character.increaseStatBy(DEX, 1)
                }
                updateViewValues()
            }
        })
        //-------------------------------------------------------------------------------------------------
    }

    fun updateViewValues() {
        character_name.text = character.name

        hp_total_val.text = character.getAttribute(HP).toString()
        hp_current_val.text = character.currentHealth.toString()

        init_val.text = character.getAttribute(INIT).toString()
        spd_val.text = character.getAttribute(SPD).toString()
        def_val.text = character.getAttribute(DEF).toString()
        dmg_val.text = if (character.equippedWeapon.subType == Melee) character.getAttribute(MDMG).toString()
                       else character.getAttribute(RDMG).toString()
        arm_val.text = character.getAttribute(ARM).toString()
        atk_val.text = character.getAttribute(ATK).toString()

        exp_total_val.text = character.expTotal.toString()
        exp_current_val.text = character.expCurrent.toString()

        per_val.text = character.getStat(PER).toString()
        chr_val.text = character.getStat(CHR).toString()
        int_val.text = character.getStat(INT).toString()
        con_val.text = character.getStat(CON).toString()
        str_val.text = character.getStat(STR).toString()
        dex_val.text = character.getStat(DEX).toString()
    }

    fun Int.toDP() : Int {
        return (this * density + 0.5).toInt()
    }

    override fun onBackPressed() {
        if (drawer_layout.isDrawerOpen(GravityCompat.START)) {
            drawer_layout.closeDrawer(GravityCompat.START)
        } else {
            super.onBackPressed()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        val id = item.itemId


        if (id == R.id.action_settings) {
            return true
        }

        return super.onOptionsItemSelected(item)
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        // Handle navigation view item clicks here.
        val id = item.itemId

        if (id == R.id.nav_camera) {
            // Handle the camera action
        } else if (id == R.id.nav_gallery) {

        } else if (id == R.id.nav_slideshow) {

        } else if (id == R.id.nav_manage) {

        } else if (id == R.id.nav_share) {

        } else if (id == R.id.nav_send) {

        }

        drawer_layout.closeDrawer(GravityCompat.START)
        return true
    }
}